File: apps/backend/src/rhesis/backend/app/outcomes.py
"""The single vocabulary for pass/fail/error across the platform.

Two axes, not one. Today's codebase stores a single free-text status name
and asks every reader to reconstruct both of these from it -- which is why
"Failed" means three unrelated things depending on where you read it. See
``playground/outcome-model/inventory.md`` for the full audit (14
vocabularies, 7 duplicated classifiers, 10 user-visible bugs) and
``playground/outcome-model/proposal.md`` for the design this implements.

``Execution`` -- did we obtain a usable observation?
    NOT_RUN, RUNNING, OK, ERROR, CANCELLED

``Verdict`` -- given it ran, did it meet its criteria?
    PASS, FAIL, INCONCLUSIVE

A verdict is only meaningful when execution is OK. That is the invariant a
single status field cannot express, and the reason "the endpoint 500'd" and
"the endpoint answered wrong" have always collapsed into one bucket.

``Outcome`` is the single derived value every layer displays and aggregates
on -- the (execution, verdict) pair projected down to one of six values.
Nothing outside this module should invent a seventh.
"""

from enum import Enum
from typing import Any, Dict, Optional, Tuple


class Execution(str, Enum):
    NOT_RUN = "not_run"
    RUNNING = "running"
    OK = "ok"
    ERROR = "error"
    CANCELLED = "cancelled"


class Verdict(str, Enum):
    PASS = "pass"
    FAIL = "fail"
    INCONCLUSIVE = "inconclusive"


class Outcome(str, Enum):
    """What every layer displays and aggregates on. Six values, no more."""

    PASS = "pass"
    FAIL = "fail"
    INCONCLUSIVE = "inconclusive"
    ERROR = "error"
    CANCELLED = "cancelled"
    PENDING = "pending"


_OUTCOME_BY_VERDICT = {
    Verdict.PASS: Outcome.PASS,
    Verdict.FAIL: Outcome.FAIL,
    Verdict.INCONCLUSIVE: Outcome.INCONCLUSIVE,
}


def outcome_of(execution: Execution, verdict: Optional[Verdict] = None) -> Outcome:
    """Project (execution, verdict) down to the one value every layer renders.

    Raises on the two combinations that would mean the caller has the model
    wrong, rather than silently picking one: a verdict is required when
    execution is OK (there is nothing else to derive an outcome from) and
    forbidden otherwise (an errored, cancelled, or unrun test has no
    verdict to report, whatever a stray value on the row might say).
    """
    if execution == Execution.OK:
        if verdict is None:
            raise ValueError("execution == OK requires a verdict")
        return _OUTCOME_BY_VERDICT[verdict]

    if verdict is not None:
        raise ValueError(f"execution == {execution.value} must not carry a verdict")

    if execution == Execution.ERROR:
        return Outcome.ERROR
    if execution == Execution.CANCELLED:
        return Outcome.CANCELLED
    return Outcome.PENDING  # NOT_RUN, RUNNING


def classify_metrics(
    metrics: Optional[Dict[str, Any]], *, http_error: bool = False
) -> Tuple[Execution, Optional[Verdict]]:
    """Classify a test result's outcome from its metrics dict.

    The single implementation of a rule that existed as seven independent,
    disagreeing copies (inventory.md's "duplication and inconsistency" #1).
    Every metric value is expected to carry ``is_successful: bool | None``
    (``None`` meaning the metric itself reported inconclusive -- see
    ``metrics/strategies/local.py``) and, on a crashed or timed-out metric,
    a truthy ``error`` key (``MetricResultBuilder.error()``/``.timeout()``).

    Precedence, most-informative first:

    1. An HTTP-level error means the endpoint never produced evaluable
       output. Metrics are meaningless even if some happen to be present
       (a stale evaluation from a retry, for instance) -- ``ERROR``.
    2. No metrics at all, or nothing metrics-shaped -- nothing was
       evaluated -- ``ERROR``.
    3. Any metric definitively failed -- ``FAIL``. A real failure is never
       masked by another metric that merely errored or was inconclusive;
       it is the strongest signal available and reviewers need to see it.
    4. Any remaining metric *crashed while evaluating* (carries an
       ``error`` key) -- ``ERROR``. Distinct from FAIL: the platform could
       not judge this metric, the system under test did nothing wrong.
    5. Any remaining metric is inconclusive (``is_successful is None``,
       no error) -- ``INCONCLUSIVE``. The metric evaluated and legitimately
       has no pass/fail verdict to give (e.g. a pure score with no
       threshold).
    6. Otherwise every metric passed -- ``PASS``.
    """
    if http_error:
        return Execution.ERROR, None

    valid = {k: v for k, v in (metrics or {}).items() if isinstance(v, dict)}
    if not valid:
        return Execution.ERROR, None

    saw_fail = False
    saw_metric_error = False
    saw_inconclusive = False

    for metric in valid.values():
        if metric.get("error"):
            saw_metric_error = True
            continue
        is_successful = metric.get("is_successful")
        if is_successful is None:
            saw_inconclusive = True
        elif not is_successful:
            saw_fail = True

    if saw_fail:
        return Execution.OK, Verdict.FAIL
    if saw_metric_error:
        return Execution.ERROR, None
    if saw_inconclusive:
        return Execution.OK, Verdict.INCONCLUSIVE
    return Execution.OK, Verdict.PASS


_STATUS_NAME_BY_OUTCOME = {
    Outcome.PASS: "Pass",
    Outcome.FAIL: "Fail",
    Outcome.ERROR: "Error",
    Outcome.INCONCLUSIVE: "Inconclusive",
}


def outcome_to_test_result_status_name(outcome: Outcome) -> str:
    """Bridge to the legacy ``status.name`` shown in review/detail UIs.

    Exists only so writers can keep populating ``status_id`` -- a
    display/review artefact, not the source of truth (see the
    ``TestResult`` model's column comment) -- without each re-inventing
    this mapping. New code that only needs the real value should read
    ``execution``/``verdict`` directly and never call this.

    Only the four outcomes a test result can actually reach at write time
    are mapped; CANCELLED and PENDING have no test-result status name
    because nothing persists a test_result row in those states.
    """
    try:
        return _STATUS_NAME_BY_OUTCOME[outcome]
    except KeyError:
        raise ValueError(f"No test-result status name for outcome {outcome.value!r}") from None


def execution_verdict_from_status_name(
    status_name: Optional[str],
) -> Tuple[Execution, Optional[Verdict]]:
    """Derive (execution, verdict) from a legacy status name directly.

    For the one write path that is allowed to supply ``status_id`` without
    ever going through ``classify_metrics`` -- a caller of
    ``POST /test_results`` may set the status explicitly with no metrics
    attached at all (see ``routers/test_result.py``). Prefer
    ``classify_metrics`` wherever metrics are actually available; this
    exists only to keep the source-of-truth columns populated when they
    aren't.
    """
    from rhesis.backend.app.constants import (
        TEST_RESULT_STATUS_FAILED,
        TEST_RESULT_STATUS_PASSED,
    )

    normalized = (status_name or "").lower()
    if normalized in TEST_RESULT_STATUS_PASSED:
        return Execution.OK, Verdict.PASS
    if normalized in TEST_RESULT_STATUS_FAILED:
        return Execution.OK, Verdict.FAIL
    if normalized == "inconclusive":
        return Execution.OK, Verdict.INCONCLUSIVE
    if normalized == "error":
        return Execution.ERROR, None
    return Execution.NOT_RUN, None

File: apps/frontend/src/constants/outcomes.ts
/**
 * Frontend outcome vocabulary mirror.
 *
 * Mirrors the two source-of-truth axes in the backend's
 * `apps/backend/src/rhesis/backend/app/outcomes.py`: `Execution` (did we
 * obtain a usable observation?) and `Verdict` (given it ran, did it meet its
 * criteria?). Both a test result and a trace carry this pair, already
 * accounting for any human review, so nothing on the client needs to
 * re-derive pass/fail from raw metrics.
 *
 * Not a classifier: `outcomeOf` is a lossless, mechanical fold with no
 * synonym lists and no fallback guessing. Keep it in sync when the backend
 * enum changes; do not add branches or heuristics here -- that logic belongs
 * server-side.
 */

export type Execution = 'not_run' | 'running' | 'ok' | 'error' | 'cancelled';
export type Verdict = 'pass' | 'fail' | 'inconclusive';

export type Outcome =
  'pass' | 'fail' | 'inconclusive' | 'error' | 'cancelled' | 'pending';

/** Anything carrying the backend's outcome pair: a test result or a trace. */
export interface HasOutcome {
  execution: Execution;
  verdict: Verdict | null;
}

/**
 * Project (execution, verdict) down to the one value a component renders.
 * Mirrors `outcomes.py`'s `outcome_of` exactly, including its invariant:
 * `verdict` only matters when `execution === 'ok'`.
 */
export function outcomeOf(
  execution: Execution,
  verdict: Verdict | null
): Outcome {
  if (execution === 'ok') {
    return verdict ?? 'pending';
  }
  if (execution === 'error') return 'error';
  if (execution === 'cancelled') return 'cancelled';
  return 'pending'; // not_run, running
}

/**
 * The display vocabulary for status chips. Four values, matching the
 * backend's `TestResultStatus` enum (constants.py) -- `Inconclusive` is a
 * real bucket, not a flavour of error: it means the metric ran and
 * legitimately has no pass/fail verdict to give (a score with no
 * threshold). The verdict grid already renders it distinctly as `scored`.
 */
export type TestResultStatus = 'Pass' | 'Fail' | 'Error' | 'Inconclusive';

const DISPLAY_STATUS: Record<Outcome, TestResultStatus> = {
  pass: 'Pass',
  fail: 'Fail',
  inconclusive: 'Inconclusive',
  error: 'Error',
  // A persisted row never actually reaches these -- nothing writes a
  // cancelled or still-pending result -- but the projection must be total.
  cancelled: 'Error',
  pending: 'Error',
};

/** The chip status for an entity carrying the backend outcome pair. */
export function displayStatusOf(entity: HasOutcome): TestResultStatus {
  return DISPLAY_STATUS[outcomeOf(entity.execution, entity.verdict)];
}

/** Past-tense label matching a display status. */
export const STATUS_LABEL: Record<TestResultStatus, string> = {
  Pass: 'Passed',
  Fail: 'Failed',
  Error: 'Error',
  Inconclusive: 'Inconclusive',
};

/**
 * "Every metric passed", for the one case the backend genuinely cannot
 * answer: a single **turn** within a multi-turn trace. A turn is finer than
 * anything the server stores an outcome for (a Trace row carries the
 * combined turn + conversation verdict, not a per-turn one), so this rule
 * has to live on the client.
 *
 * Use this ONLY for per-turn or per-metric-group summaries. For anything
 * that has an `execution`/`verdict` pair -- a test result, a trace -- use
 * `displayStatusOf` instead; re-deriving there is what this whole module
 * exists to prevent.
 *
 * Note this is deliberately weaker than the backend's `classify_metrics`:
 * it has no notion of a crashed metric or an inconclusive one, because a
 * turn summary has nowhere to render those. Empty means false, not "pass
 * by default".
 */
export function allMetricsPassed(
  metrics: Array<{ is_successful?: boolean | null }>
): boolean {
  return metrics.length > 0 && metrics.every(m => m.is_successful === true);
}

// ---------------------------------------------------------------------------
// Pass rate
// ---------------------------------------------------------------------------

/**
 * The one pass-rate formula, as a percentage (0-100), or null when there is
 * nothing to rate.
 *
 * Mirrors the backend's `pass_rate_of_executed`: the denominator is what
 * actually resolved to pass or fail. An errored, inconclusive, or unrun
 * subject is excluded rather than counted as a failure, so the rate answers
 * "of the things we could judge, how many passed" instead of silently
 * changing meaning with the number of unrun items.
 *
 * Returns null (not 0) for an empty denominator so callers must decide how
 * "no data" renders instead of showing a misleading 0%.
 */
export function passRate(passed: number, failed: number): number | null {
  const resolved = passed + failed;
  if (resolved <= 0) return null;
  return (passed / resolved) * 100;
}

// ---------------------------------------------------------------------------
// Pass-rate banding
// ---------------------------------------------------------------------------

export type ReviewBand = 'ok' | 'watch' | 'review';

export interface ReviewBandInfo {
  band: ReviewBand;
  label: string;
  colorKey: 'success' | 'warning' | 'error';
}

/**
 * The single pass-rate colour scale. Previously three scales disagreed
 * (100/70 here, 70/40 on the insights summary bar, 80 on the history tab),
 * two of them on the same page, so the same rate could be green in one
 * widget and red in the next.
 */
export function getReviewBand(passRateValue: number): ReviewBandInfo {
  if (passRateValue >= 100) {
    return { band: 'ok', label: 'OK', colorKey: 'success' };
  }
  if (passRateValue >= 70) {
    return { band: 'watch', label: 'Watch', colorKey: 'warning' };
  }
  return { band: 'review', label: 'Needs Review', colorKey: 'error' };
}
